engine/src/mac-core.mm

```diff
@@ -803,16 +803,18 @@ bool MCPlatformWaitForEvent(double p_duration, bool p_blocking)
 										inMode: p_blocking ? NSEventTrackingRunLoopMode : NSDefaultRunLoopMode
 									   dequeue: NO];
 		
+		// Fetch the most deeply-nested modal session to run
+		MCModalSession *t_session = s_modal_sessions[s_modal_sessions.Size() - 1];
```

Comment:
So I might be missing something - but couldn't you just record the session index here - i.e. s_modal_sessions.size() - 1?

As we only push/pop sessions onto the modal_sessions array there can't be a lifetime issue here (as that would still exist with using pointers)...

Indeed, I'm guessing the problem is that the run_depth as previously calculated can be wrong if runModalSession: launches another modal session recursively.

Reply:
My thinking is that when setting / unsetting session.is_running, if I use an index into the array what happens if during the modal session that session ends and another is begun? or just if the session ends then the index points off the end of the array. The MCModalSession object pointed to persists until cleanup so the pointer should still be valid up to that point